src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java

```diff
@@ -153,13 +158,28 @@ public TaskResult call() {
                 recordProcessor.shutdown(shutdownInput);
                 ExtendedSequenceNumber lastCheckpointValue = recordProcessorCheckpointer.getLastCheckpointValue();
 
+                final boolean successfullyCheckpointedShardEnd = lastCheckpointValue.equals(ExtendedSequenceNumber.SHARD_END);
+
                 if (localReason == ShutdownReason.TERMINATE) {
-                    if ((lastCheckpointValue == null)
-                            || (!lastCheckpointValue.equals(ExtendedSequenceNumber.SHARD_END))) {
+                    if ((lastCheckpointValue == null) || (!successfullyCheckpointedShardEnd)) {
                         throw new IllegalArgumentException("Application didn't checkpoint at end of shard "
                                 + shardInfo.getShardId() + ". Application must checkpoint upon shutdown. " +
                                 "See IRecordProcessor.shutdown javadocs for more information.");
                     }
+
+                    // Check if either the shard end ddb persist is successful or
+                    // if childshards is empty. When child shards is empty then either it is due to
+                    // completed shard being reprocessed or we got RNF from service.
+                    // For these cases enqueue the lease for deletion.
+                    if (successfullyCheckpointedShardEnd || CollectionUtils.isNullOrEmpty(childShards)) {
+                        final KinesisClientLease currentLease = leaseCoordinator.getCurrentlyHeldLease(shardInfo.getShardId());
+                        final LeasePendingDeletion leasePendingDeletion = new LeasePendingDeletion(currentLease, shardInfo);
+
+                        if (!leaseCleanupManager.isEnqueuedForDeletion(leasePendingDeletion)) {
+                            leaseCleanupManager.enqueueForDeletion(leasePendingDeletion);
```

Comment:
Add a comment here for Jessica to do shard end checkpointing here.

Reply:
Will add